### DISCUSSION TITLE
CLI: Improvements to command line arguments. Add `-mode`

### DISCUSSION THREAD

**Discussion by jgrg** [duckdb/duckdb#12831] [Ideas]
I think it would make sense to have a `-mode <NAME>` option, to choose any of the `.mode` dot command options, rather than having a separate option for each. We currently have:
|   Option     |            Effect              |      Note      |
|--------------|--------------------------------|----------------|
| `-ascii`     | Set output mode to ascii       |                |
| `-box`       | Set output mode to box         |                |
| `-column`    | Set output mode to column      |                |
| `-csv`       | Set output mode to csv         |                |
| `-html`      | Set output mode to HTML        |                |
| `-json`      | Set output mode to json        |                |
| `-jsonlines` | Set output mode to nd-json     | Not documented |
| `-latex`     | LaTeX tabular environment code | Not documented |
| `-line`      | Set output mode to line        |                |
| `-list`      | Set output mode to list        |                |
| `-markdown`  | Set output mode to markdown    |                |
| `-quote`     | Set output mode to quote       |                |
| `-table`     | Set output mode to table       |                |

Missing from command line:

|  .mode Name  |             Meaning             |                         Note                            |
|--------------|---------------------------------|---------------------------------------------------------|
| duckbox      | Tables with extensive features  | Useful if there is a different `.mode` in `~/.duckdbrc` |
| insert       | SQL insert statements for TABLE |                                                         |
| tabs         | Tab-separated values            |                                                         |
| tcl          | TCL list elements               |                                                         |
| trash        | No output                       | Might never be used?                                    |

### Cannot set `tabs` mode

I cannot set `tabs` mode to make TSV output. It isn't one of the command line options, and `-list -separator "\t"` produces literal `\t` as column separators rather than tab characters.

### Output format options ignored

I do find it useful being able to take input from a UNIX pipe into duckdb, _e.g._
```sh
my_command | duckdb -c "FROM read_json('/dev/stdin') WHERE url IS NOT NULL"
```
This prints to the terminal in `duckbox` output mode. Maybe I then decide I want to save the output in CSV file. This doesn't work, it still outputs in `duckbox`:
```sh
my_command | duckdb -c "FROM read_json('/dev/stdin') WHERE url IS NOT NULL" -csv > my_file.csv
```
because the order of arguments appears to be important, and if `-csv` (or any other output format option) is after the `-c <SQL>` it is ignored. This does work:
```sh
my_command | duckdb -csv -c "FROM read_json('/dev/stdin') WHERE url IS NOT NULL" > my_file.csv
```

**Comment by Tishj** [comment]
Feel free to make a PR, good place to start is probably here:
<https://github.com/duckdb/duckdb/blob/b43a234a67d038c700bcd4ed0db8aafcb0d6a6b2/tools/shell/shell.c#L20002>

**Reply by jgrg** [reply]
Ha! Reading the source code I realised I can get TSV output with:
```sh
my_cmd | duckdb -c ".mode tabs" -c "FROM read_json('/dev/stdin') WHERE url IS NOT NULL" > my_file.tsv
```
I'm not a C programmer, so I'm not likely to make a pull request. Looking at the code I don't see why a `-csv` after  a `-c` is ignored…

**Reply by jgrg** [reply]
Oh, and `shell.c` is a 20,000 line file which begins:
```c
/* DO NOT EDIT!
** This file is automatically generated by the script in the canonical
** SQLite source tree at tool/mkshellc.tcl.  That script combines source
** code from various constituent source files of SQLite into this single
** "shell.c" file used to implement the SQLite command-line shell.
```

**Reply by Tishj** [reply]
The shell is originally from sqlite, it has since been modified quite a bit

**Reply by jgrg** [reply]
The file with **DO NOT EDIT** at the top has been edited?
